plugins/transcend/transcend-nvme.c

```diff
@@ -5,82 +5,396 @@
 #include <stdlib.h>
 #include <unistd.h>
 #include <inttypes.h>
+#include <string.h>
+#include <time.h>
 
 #include "nvme.h"
 #include "libnvme.h"
 #include "plugin.h"
+#include "transcend-nvme.h"
+#include "common.h"
 
 #define CREATE_CMD
-#include "transcend-nvme.h"
 
-static const __u32 OP_BAD_BLOCK = 0xc2;
+#define PLP_RECORD_PATH		"PLPRec.txt"
+#define PLP_DATA_EXPIRED_TIME	20
+
+
+/* Bad block command parameters */
+static const __u32 OP_BAD_BLOCK  = 0xc2;
 static const __u32 DW10_BAD_BLOCK = 0x400;
 static const __u32 DW12_BAD_BLOCK = 0x5a;
 
-static int getHealthValue(int argc, char **argv, struct command *cmd, struct plugin *plugin)
+/* Health command parameters */
+static const __u32 OP_HEALTH	= 0xc2;
+static const __u32 DW10_HEALTH	= 0x01;
+static const __u32 DW12_HEALTH	= 0xb3;
+
+/* ID command parameters */
+static const __u32 OP_ID	= 0x06;
+static const __u32 DW10_ID	= 0x01;
+
+/* PLP constants */
+static const int i_dis		= 20;
+static const double full_value	= 170;
+
+enum plp_error_code {
+	PLP_ERROR_NO_MATCH	= -1,
+	PLP_ERROR_DATA_EXPIRED	= -2,
+};
+
+static const char *string_list[] = {
+	"UTE210T",
+	"MTE712P",
+	"MTE560P",
+	"MTE662P",
+	"MTE662P-I",
+	"MTS970P",
+	"MTS952P",
+	"MTS952P-I",
+	"MTS570P",
+	"MTS400P",
+	"SSD910T",
+	"SSD470P",
+	"SSD470P-I",
+	"SSD452P",
+	"SSD452P-I",
+	"SSD420P",
+	"MSA470P",
+	"MSA452P",
+};
+
+static const int list_size = ARRAY_SIZE(string_list);
+
+static int get_plp_health(int argc, char **argv, struct command *cmd,
+			  struct plugin *plugin);
+static int read_useful_plp_value(const char *device);
+static void record_plp_value(const char *device, int value, bool is_replace);
+
+static int get_health_value(int argc, char **argv, struct command *cmd,
+			    struct plugin *plugin)
 {
 	struct nvme_smart_log smart_log;
-	char *desc = "Get nvme health percentage.";
-	int  percent_used = 0, healthvalue = 0;
 	struct nvme_dev *dev;
-	int result;
-
+	char *desc = "Get nvme health percentage.";
+	int percent_used = 0;
+	int health_value = 0;
+	int ret;
+	
 	OPT_ARGS(opts) = {
 		OPT_END()
 	};
 
-	result = parse_and_open(&dev, argc, argv, desc, opts);
-	if (result) {
-		printf("\nDevice not found\n");
-		return -1;
+	ret = parse_and_open(&dev, argc, argv, desc, opts);
+	if (ret) {
+		pr_err("Device not found\n");
```

Comment:
Seems better to use `fprintf(stderr, ...)` and `printf()` instead of the `pr_err` and `pr_info` since only this file using the `pr_***()` funcitons.